bin/validate_bigquery

```diff
@@ -0,0 +1,167 @@
+import json
+import difflib
+import sys
+
+BASE_DIR = Path("/app").resolve()
+
+
+@click.group()
+def validate():
+    pass
+
+
+def compute_compact_columns(document):
+    def traverse(prefix, columns):
+        res = []
+        for node in columns:
+            name = node["name"] + (".[]" if node["mode"] == "REPEATED" else "")
+            dtype = node["type"]
+            if dtype == "RECORD":
+                res += traverse(f"{prefix}.{name}", node["fields"])
+            else:
+                res += [f"{prefix}.{name} {dtype}"]
+        return res
+
+    res = traverse("root", document)
+    return sorted(res)
+
+
+@validate.command()
+@click.option("--head", default="HEAD")
+@click.option(
+    "--repository",
+    type=click.Path(exists=True, file_okay=False),
+    default=BASE_DIR / "mozilla-pipeline-schemas",
+)
+@click.option(
+    "--artifact",
+    type=click.Path(file_okay=False),
+    default=BASE_DIR / "validate_schema_evolution",
+)
+def copy(head, repository, artifact):
+    """Copy BigQuery schemas to a directory as an intermediary step for schema
+    evolution checks."""
+    src = Path(repository)
+    repo = Repo(repository)
+    dst = Path(artifact) / repo.rev_parse(head).name_rev.replace(" ", "_")
+    dst.mkdir(parents=True, exist_ok=True)
+    schemas = sorted(src.glob("**/*.bq"))
+    if not schemas:
+        raise click.ClickException("no schemas found")
+    for path in schemas:
+        namespace = path.parts[-3]
+        doc = path.parts[-1]
+        qualified = f"{namespace}.{doc}"
+        click.echo(qualified)
+        copyfile(path, dst / qualified)
+
+        # also generate something easy to diff
+        cols = compute_compact_columns(json.loads(path.read_text()))
+        compact_filename = ".".join(qualified.split(".")[:-1]) + ".txt"
+        (dst / compact_filename).write_text("\n".join(cols))
+
+
+def check_evolution(base, head, verbose=False):
+    def nop(*args, **kwargs):
+        pass
+
+    log = print if verbose else nop
+
+    a, b = set(base), set(head)
+    status = 0
+    # error condition
+    base_only = a - b
+    if len(base_only) > 0:
+        log("items removed from the base")
+        log("\n".join([f"-{x}" for x in base_only]))
+        log("")
+        # set the status
+        status = 1
+
+    # informative only
+    head_only = b - a
+    if len(head_only) > 0:
+        log("items added to the base")
+        log("\n".join([f"+{x}" for x in head_only]))
+        log("")
+    return status
+
+
+@validate.command()
+@click.option("--head", type=str, default="local-working-branch")
+@click.option("--base", type=str, default="generated-schemas")
+@click.option(
+    "--repository",
+    type=click.Path(exists=True, file_okay=False),
+    default=BASE_DIR / "mozilla-pipeline-schemas",
+)
+@click.option(
+    "--artifact",
+    type=click.Path(file_okay=False),
+    default=BASE_DIR / "validate_schema_evolution",
+)
+def local(head, base, repository, artifact):
```

Comment:
Should this be the validate function?

Reply:
This naming is to contrast with another function I intended ("remote") which applies the schemas to bigquery, but this might be misnamed a this point in time.

Reply:
Got it, let me know what you think about making it all comparison of local branches.